module/bdev/ocf/ctx.c

```diff
@@ -533,6 +539,195 @@ vbdev_ocf_ctx_log_printf(ocf_logger_t logger, ocf_logger_lvl_t lvl,
+	if (pmeta->fd < 0) {
+		SPDK_ERRLOG("Can't open SHM %s\n", pmeta->name);
+		return NULL;
+	}
+
+	if (ctx->force) {
+		SPDK_NOTICELOG("Using force, truncating SHM %s\n",
+				pmeta->name);
+		if (ftruncate(pmeta->fd, 0) < 0) {
+			SPDK_ERRLOG("Failed to truncate\n");
+			goto err_stat;
+		}
+
+	} else if (fstat(pmeta->fd, &stat_buf) < 0) {
+		SPDK_ERRLOG("Can't stat SHM %s\n", pmeta->name);
+		goto err_stat;
+	}
+
+	if (!ctx->force && stat_buf.st_size != 0) {
+		if (ctx->create) {
+			SPDK_ERRLOG("SHM %s found, but no force specified!\n",
+					pmeta->name);
+			goto err_stat;
+		}
+
+		*load = true;
+		SPDK_NOTICELOG("Loading from SHM\n");
+		if (stat_buf.st_size < size) {
+			if (ftruncate(pmeta->fd, size) < 0) {
+				SPDK_ERRLOG("Failed to extend SHM\n");
+				goto err_stat;
+			}
+			pmeta->size = size;
+		} else if (stat_buf.st_size == size) {
+			pmeta->size = size;
+		} else {
+			SPDK_NOTICELOG("Refusing to shrink SHM\n");
+			pmeta->size = stat_buf.st_size;
+		}
+	} else {
+		SPDK_NOTICELOG("Set SHM size\n");
+		if (ftruncate(pmeta->fd, size) < 0) {
+			SPDK_ERRLOG("Failed to truncate\n");
+			goto err_stat;
+		}
+
+		pmeta->size = size;
+	}
+
+	shm = mmap(NULL, pmeta->size, PROT_READ | PROT_WRITE, MAP_SHARED, pmeta->fd, 0);
+	if (shm == MAP_FAILED) {
+		SPDK_ERRLOG("Failed to map shm\n");
+		goto err_stat;
+	}
+
+	pmeta->data = shm;
+	shm_sb = shm;
+
+	if (mlock(shm, pmeta->size)) {
+		SPDK_ERRLOG("Failed to mlock\n");
+
+		goto err_mlock;
+	}
+
+
+	if (!*load) {
+		shm_sb->segments[0].size = sizeof(struct shm_superblock);
+		shm_sb->segments[0].offset = 0;
+		shm_sb->segments[0].id = -1;
+		shm_sb->segments[0].valid = true;
+	}
+
+	return pmeta;
+
+err_mlock:
+	munlock(pmeta->data, pmeta->size);
+err_stat:
+	close(pmeta->fd);
+	shm_unlink(pmeta->name);
+
+	return NULL;
+}
+
+static int
+vbdev_ocf_persistent_meta_deinit(struct ocf_persistent_meta_zone *zone)
+{
+	/*
+	munmap(zone->data, zone->size);
+	close(zone->fd);
+	shm_unlink(zone->name);
+	*/
+
+	return 0;
+}
+
+static void *
+vbdev_ocf_persistent_meta_alloc(struct ocf_persistent_meta_zone *zone, size_t size,
+		int alloc_id, bool *load)
+{
+	int i;
+	size_t sum = 0;
+	struct shm_superblock *shm_sb = zone->data;
+	struct ocf_cache_persistent_meta_segment_descriptor *desc =
+		shm_sb->segments;
+
+	alloc_id++;
+	*load = false;
+
+	for (i = 0; i < MAX_SEGMENTS; i++) {
+		/* assume no fragmentation */
+		if (!desc[i].valid)
+			break;
+
+		sum += desc[i].size;
+
+		/* assume no size change */
+		if (desc[i].id == alloc_id) {
+			*load = true;
+			return zone->data + desc[i].offset;
+		}
+	}
+
+	if (i == MAX_SEGMENTS)
+		return NULL;
+
+	 /* TODO: align */
+
+	if (sum + size > zone->size)
+		return NULL;
+
+	desc[i].id = alloc_id;
+	desc[i].offset = sum;
+	desc[i].size = size;
+	desc[i].valid = true;
+
+	return zone->data + desc[i].offset;
+}
+
+static int
+vbdev_ocf_persistent_meta_free(struct ocf_persistent_meta_zone *zone, int alloc_id,
+		void *ptr)
+{
```

Comment:
Should this invalidate the appropriate ocf_cache_persistent_meta_segment_descriptor ?

Reply:
this is intentionally left blank for now